src/tools/rbd_mirror/image_map/ImageAction.h

```diff
@@ -0,0 +1,122 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#ifndef CEPH_RBD_MIRROR_IMAGE_MAP_IMAGE_ACTION_H
+#define CEPH_RBD_MIRROR_IMAGE_MAP_IMAGE_ACTION_H
+
+#include <map>
+#include <list>
+#include <string>
+
+#include "Types.h"
+#include "common/Mutex.h"
+
+namespace rbd {
+namespace mirror {
+namespace image_map {
+
+class ImageAction {
+public:
+  ImageAction();
+
+  enum ActionType {
+    ACTION_TYPE_ADD = 0,
+    ACTION_TYPE_REMOVE,
+    ACTION_TYPE_SHUFFLE,
+  };
+
+  enum SubActionType {
+    SUB_ACTION_TYPE_INVALID = 0,
+    SUB_ACTION_TYPE_MAP,
+    SUB_ACTION_TYPE_UNMAP,
+    SUB_ACTION_TYPE_PURGE,
+    SUB_ACTION_TYPE_COMPLETE,
+  };
+
+  struct Action {
+    ActionType action_type;
+
+    Context *on_map;
+    Context *on_unmap;
+    Context *on_purge;
+    Context *on_finish;
+
+    SubActionType sub_action_type; // currently executing sub action
+
+    Action(ActionType action_type)
+      : action_type(action_type),
+        on_map(nullptr),
+        on_unmap(nullptr),
+        on_purge(nullptr),
+        on_finish(nullptr),
+        sub_action_type(SUB_ACTION_TYPE_INVALID) {
+    }
+  };
+
+  void init_state(const std::string &global_image_id);
+  void release_state(const std::string &global_image_id);
+
+  bool queue_action(const std::string &global_image_id, const Action &action);
+
+  bool actions_pending(const std::string &global_image_id);
+
+  bool start_action(const std::string &global_image_id);
+  bool finish_action(const std::string &global_image_id);
+
+  SubActionType sub_action_start(const std::string &global_image_id);
+  bool sub_action_finish(const std::string &global_image_id);
+
+private:
+  typedef std::list<Action> Actions;
+
+  enum State {
+    STATE_UNASSIGNED = 0,
+    STATE_ASSOCIATED,
+    STATE_DISASSOCIATED,
+    STATE_PURGING,
+    STATE_SHUFFLING,
+  };
+
+  struct Transition {
+    SubActionType sub_action_type;
+    State next_state;
+  };
+
+  struct TransitionTable {
+    ActionType action_type;
+    State current_state;
+    Transition transition;
+  };
+  static const TransitionTable transition_table[];
+
+  struct ActionState {
+    Actions actions;                // list of pending actions
+    std::pair<State, State> states; // current,next state
+
+    ActionState()
+      : states(std::make_pair(STATE_UNASSIGNED, STATE_UNASSIGNED)) {
+    }
+  };
+
+  Mutex m_lock;
```

Comment:
Is the lock required? Looks like the ```Policy::m_map_lock``` is always locked when using the ```ImageAction``` instance

Reply:
Technically, its not required. But I'm thinking if at all locking in `ImageAction` is required in future. Also, this lock is never contented as of now (due to all access is under `Policy::m_map_lock`), so its pretty much lightweight.

Reply:
Personally -- I'd yank it since ```ImageAction``` should always be used under the image map's lock